File: training-data/python/factory_pattern.py
from abc import ABC, abstractmethod


class Shape(ABC):
    @abstractmethod
    def area(self) -> float: ...


class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self) -> float:
        return 3.14159 * self.radius ** 2


class Square(Shape):
    def __init__(self, side):
        self.side = side

    def area(self) -> float:
        return self.side ** 2


def shape_factory(kind: str, **kwargs) -> Shape:
    shapes = {"circle": Circle, "square": Square}
    if kind not in shapes:
        raise ValueError(f"unknown shape: {kind}")
    return shapes[kind](**kwargs)


if __name__ == "__main__":
    print(shape_factory("circle", radius=2).area())
    print(shape_factory("square", side=3).area())

File: training-data/typescript/factory_pattern.ts
interface Shape {
  area(): number;
}

class Circle implements Shape {
  constructor(private radius: number) {}
  area(): number {
    return Math.PI * this.radius ** 2;
  }
}

class Square implements Shape {
  constructor(private side: number) {}
  area(): number {
    return this.side ** 2;
  }
}

type ShapeKind = 'circle' | 'square';

function shapeFactory(kind: ShapeKind, options: { radius?: number; side?: number }): Shape {
  if (kind === 'circle') return new Circle(options.radius!);
  if (kind === 'square') return new Square(options.side!);
  throw new Error(`unknown shape: ${kind}`);
}

console.log(shapeFactory('circle', { radius: 2 }).area());
console.log(shapeFactory('square', { side: 3 }).area());
